joblib/numpy_pickle.py

```diff
@@ -1,330 +1,335 @@
+    """An object to be persisted instead of numpy arrays.
 
-def hex_str(an_int):
-    """Converts an int to an hexadecimal string
+    This object is used to hack into the pickle machinery and read numpy
+    array data from our custom persistence format.
+    More precisely, this object is used for:
+    * carrying the information of the persisted array: subclass, shape, order,
+    dtype. Those ndarray metadata are used to correctly reconstruct the array
+    with low level numpy functions.
+    * determining if memmap is allowed on the array.
+    * reading the array bytes from a file.
+    * reading the array using memorymap from a file.
+    * writing the array bytes to a file.
+
+    Attributes
+    ----------
+    subclass: numpy.ndarray subclass
+        Determine the subclass of the wrapped array.
+    shape: numpy.ndarray shape
+        Determine the shape of the wrapped array.
+    order: {'C', 'F'}
+        Determine the order of wrapped array data. 'C' is for C order, 'F' is
+        for fortran order.
+    dtype: numpy.ndarray dtype
+        Determine the data type of the wrapped array.
+    allow_mmap: bool
+        Determine if memory mapping is allowed on the wrapped array.
+        Default: False.
     """
-    return '{0:#x}'.format(an_int)
-
 
-_MEGA = 2 ** 20
+    def __init__(self, subclass, shape, order, dtype, allow_mmap=False):
+        """Constructor. Store the useful information for later."""
+        self.subclass = subclass
+        self.shape = shape
+        self.order = order
+        self.dtype = dtype
+        self.allow_mmap = allow_mmap
 
-# Compressed pickle header format: _ZFILE_PREFIX followed by _MAX_LEN
-# bytes which contains the length of the zlib compressed data as an
-# hexadecimal string. For example: 'ZF0x139              '
-_ZFILE_PREFIX = asbytes('ZF')
-_MAX_LEN = len(hex_str(2 ** 64))
+    def write_array(self, array, pickler):
+        """Write array bytes to pickler file handle.
 
+        This function is an adaptation of the numpy write_array function
+        available in version 1.10.1 in numpy/lib/format.py.
+        """
+        # Set buffer size to 16 MiB to hide the Python loop overhead.
+        buffersize = max(16 * 1024 ** 2 // array.itemsize, 1)
+        if array.dtype.hasobject:
+            # We contain Python objects so we cannot write out the data
+            # directly. Instead, we will pickle it out with version 2 of the
+            # pickle protocol.
+            pickle.dump(array, pickler.file_handle, protocol=2)
+        else:
+            for chunk in pickler.np.nditer(array,
+                                           flags=['external_loop',
+                                                  'buffered',
+                                                  'zerosize_ok'],
+                                           buffersize=buffersize,
+                                           order=self.order):
+                pickler.file_handle.write(chunk.tostring('C'))
```

Comment:
Question: is it possible to use the memoryview directly?  This might avoid a copy

```python
In [1]: import numpy as np

In [2]: x = np.ones(5)

In [3]: x.data
Out[3]: <memory at 0x7f201c81fe58>

In [4]: with open('foo.pkl', 'wb') as f:
    f.write(x.data)
```